### ISSUE TITLE
91st Meridian - Create writing collection container + writing content type

### ISSUE THREAD

**Issue by pyrello** [uiowa/uiowa#5601]
# Motivation

91st Meridian is a periodical literary magazine. As such, it is organized into issues, with [each issue](https://iwp.uiowa.edu/91st/vol1-num1) containing [multiple individual pieces of writing](https://iwp.uiowa.edu/91st/vol1-num1/too-little-seen). In the current version, there are two content types: 91st Meridian Content and 91st Meridian Issue Homepage. I believe this idea of having an entity represent the collection and a different entity to represent the individual pieces of writing is a good way to organize it.

After taking a look at iowareview.org, I think that represents a good path forward for this work.

# Proposed solution

- [x] Add a Periodical/Journal feature split
- [x] Add an Issue content type:
  - [x] Volume
  - [x] Issue Number
  - [x] Season/Year
  - [x] Issue Release Date
  - [x] Table of Contents - _Use body field?_
  - [x] Cover Image (`field_image`)
  - [ ] Description/Teaser
  - [x] Editor's Note
  - [x] Masthead
  - [x] Contributors' Notes
- [x] Add an Issue article content type
  - [x] Issue - ER field (cardinality = 1) that can reference the Issue content type
  - [x] Body
  - [x] Author (text or ER?)
  - [x] Author Bio
  - [ ] Genre
- [ ] Both content types will need a custom path alias pattern
- [ ] Check https://wiki.uiowa.edu/display/drupaldev/Content+type+creation+notes to see other steps that may not be documented here for new content types

<details>
<summary>Original proposed solution</summary>
This is currently based on the investigation at https://github.com/uiowa/uiowa/issues/3868#issuecomment-1222440670

- Add a Periodical (?) feature split
- Enable Entity Hierarchy + submodules
- Add a issue/collection content type
- Add a writing/article content type
- Add all new config to the feature split and export config
</details>

**Comment by mjoneill** [comment]
We have similar needs (presumably) with all our campus journals.  Law journals are coming up this fall...

I don't think this is required for that use case, but it might be a more elegant solution for collecting related content, depending on how you go with it...

**Comment by joewhitsitt** [comment]
I took an early peek at the entity_hierarchy and sub-modules. Based on the requirements of this issue I am not sure I see a big benefit of this module over just using a custom menu system tied to these two content types except to prevent a collection from being nested under a piece of writing.

This module is a weighted reference field that provides a children tab to visually reorder one level at a time. Two sub-modules are necessary to generate the breadcrumb and a menu around this relationship. From the issue queue, it seems like operations  around this functionality and the menu/breadcrumb are delayed or need to be rebuilt manually. I was able to cause the menu to go out of sync by manually managing the menu and then again later by deleting one of the writings.

Similarly for Admissions we did a local event tied to an event series which displays the items on the event series sorted by date. In place of date for this situation, we could set this up with the weight module or draggableviews.

I do see this as a potential solution for less SiteNow sites down the road by allowing a "conferences.uiowa.edu" type website with sub-sites where we need to control editing with workbench access.

**Comment by pyrello** [comment]
We had a discussion about this and here are the ideas that we came up with:
- Writing content type just uses a normal ER field pointing to the writing collection and is set up with IEF to allow on the fly creation of writing collections
- Create a new menu called Periodicals
- Allow writing collections and writing content types to create menu items in the Periodical menu
- Place the Periodical menu block on the writing collection LB template and set visibility to show children
- Place the Periodical menu block on the writing LB template and set visibility to show siblings
- Custom path alias pattern
- Bonus: Figure out how to set the menu parent in the node form based on the collection that is set on the writing content

**Comment by pyrello** [comment]
As I've started to dig into 91st Meridian a bit, there are a lot of inconsistencies with how each issue is built. Also, it uses a bespoke format for the table of contents that doesn't lend itself well to a standard solution. I have a meeting set up to chat with someone at IWP with previous experience in literary magazines to understand better what should or should not be there. After that, I'll be in a better position to try to understand how this can fit in with a feature that might be used by other sites or whether it might need to just be custom.

**Comment by pyrello** [comment]
This has been paused while we were figuring out how we could potentially wrangle the existing 91st Meridian content into a more standard package that could be shared with other periodical websites. Here are some examples I have been looking at for inspiration:
- https://iowareview.org/
- https://exchanges.uiowa.edu/
- https://www.vqronline.org/issues/92/2/spring-2016
- https://www.nplusonemag.com/magazine/issue-43/

**Comment by pyrello** [comment]
Iowa Review uses a text area for the table of contents rather than using a menu. The onus is on the content editors to ensure that the treatment is consistent, but it is flexible and allows for different ways of structuring the table of contents. The reason for doing this on Iowa Review is that not every article is published online, so this prevents the site from needing a way to define the articles that are not published.

This happens to be consistent with how 91st Meridian is doing things currently.

**Comment by pyrello** [comment]
Update IS to reflect the removal of integrating with a new menu to handle the Table of Contents since we are using a dedicated text area field for that.

**Comment by joewhitsitt** [comment]
Is it too late to consider https://schema.org/Periodical https://schema.org/PublicationIssue as part of this using the Schema.org blueprints module? While probing https://github.com/uiowa/uiowa/issues/5810, it just makes me think that we should consider a "schema-first" approach going forward, especially with feature types that might be used on more than one site.

**Comment by pyrello** [comment]
> Is it too late to consider https://schema.org/Periodical https://schema.org/PublicationIssue as part of this using the Schema.org blueprints module? While probing #5810, it just makes me think that we should consider a "schema-first" approach going forward, especially with feature types that might be used on more than one site.

Not too late. From your perspective, what would we need to change about the proposed solution to accomplish this?

**Comment by joewhitsitt** [comment]
I'd have to take a closer look, but to me it sounds like there are more layers to account for with the PublicationIssue referencing a PublicationVolume?

Periodical > PublicationVolume > PublicationIssue > Issue Article

So the proposed solution would change based on those relationships and that the blueprints module would be the starting point and so https://github.com/uiowa/uiowa/pull/5869 would keep the modules available in default

**Comment by joewhitsitt** [comment]
If work on this begins before https://github.com/uiowa/uiowa/pull/5869 is merged, make sure the module dependencies and settings for blueprints are declared the same way between local and default config.

**Comment by briand44** [comment]
Theming not in scope for this slice. Trying to unblock content migration work.

**Comment by pyrello** [comment]
We just huddled on this and spent some time looking into the implications of @joewhitsitt's [comment](https://github.com/uiowa/uiowa/issues/5601#issuecomment-1284457631). Here is what we learned:
- There are some different examples of how to construct the SDO record for a PublicationIssue and PublicationArticle: https://schema.org/PublicationIssue#eg-0398. We are still in the process of grokking how that translates into how we will want to set up our records.
- In the cases of both 91st Meridian and Iowa Review, the volume number is just a property of the issue. It seems like it might be overkill to have a content entity to represent it. We discussed using a term reference field and vocabulary to handle it because it would allow us to easily list issues that belong to a particular volume.
- In most cases it seems like there is a 1:1 relationship between a site and the publication on it, so we might be able to hard-code that based on the site name. OTOH, @quamsta indicated that https://exchanges.uiowa.edu/ has several sub-journals. Probably some more conversation is necessary to understand how that case fits in to what we are doing here. If it becomes necessary to have multiple publications, then we might also consider using a vocab to represent those.
- There is a Schema.org sub-module for enabling the functionality for taxonomy that we would need to turn on for the above situations.
- It is possible to add a SDO definition to an existing content type via Fields > Schema.org tab.

**Comment by e-marie-w** [comment]
Interesting discovery, a field can directly reference a schema.org mapping or schema.org mapping type. Which I don't think is what we want but indirectly made me more confused about schema, so I want to check my understanding.

So there's the different schema.org types - Publication Volume, Publication Issue, etc. And they each have type-specific properties. And there's no real overlap or connection between the different types and their properties.

But for our purposes we are connecting some of the information. So, for example, I add Publication Volume as a taxonomy vocabulary, and Publication Issue as a content type. And then I add a field to the content type that references the taxonomy terms, since the issue is part of a larger volume. And, at first, I thought I could map the volumeNumber property to that new field referencing the Publication Volume. But I can't, since it's a property of Publication Volume and so can only map to a field within that vocabulary. 

So, I guess my questions, besides is anything wrong above, are: when referencing an entity is the schema for that entity also referenced? And does it matter - we have the issue and volume linked, and the schemas are defined for both, so does it matter there isn't, like, the Issue isPartOf Volume like in their examples?

**Comment by joewhitsitt** [comment]
So my understanding is that if the both sides of the relationship are mapped to schema.org properties through field types then the overall schema json+ld output should include some if not all of those details. We noticed that with event and person. Despite the relationship to a performer did not show properly in the event schema markup. it wasn't until that i mapped the person content type and some of it's fields to the schema.org person type that those performers started showing on the event.

I would see how far you can get without any custom code and see if it validates at https://validator.schema.org/ by copy/pasting your local output. We had to fill some gaps for the event with custom alters and we can do so here as well.

**Comment by e-marie-w** [comment]
How do I get the local output? If I view page source I can find a <script type="application/ld+json"> section like the examples on schema.org. It won't validate since it's code but I'm guessing it's another way to check?

**Comment by joewhitsitt** [comment]
It should be nested in the `<head>` tag

<img width="495" alt="Screen Shot 2022-11-07 at 10 22 06 AM" src="https://user-images.githubusercontent.com/4663676/200361485-c5685f0b-36f2-4272-8cc4-5056e71847ab.png">

There should be two of them if you are using schemadotorg blueprints. Just copy the inner json and paste in the "code snippet" option on that validator

**Comment by e-marie-w** [comment]
I'm doing something wrong, I paste in their examples and still get 'There was an error validating your markup'. I skip the <script> tag, then paste what's in the curly brackets (tried with and without including the brackets). Also tried different examples and combinations, still only get errors. Any ideas? 

Otherwise I already can tell not all properties are showing and will work on that.

**Comment by joewhitsitt** [comment]
>Any ideas?

Can you paste here what you were trying paste there so we can look at it? Otherwise, let's find a time where you can huddle.

Update: huddled and showed off the browser inspector > find > "json" > copy element

**Comment by e-marie-w** [comment]
I'm wondering if we want to add the issue release date and description/teaser fields to the Publication Issue at this time, as they're Iowa Review specific. The issue release date doesn't display on the issue, but it might be how they sort the recent issues page. Then again, couldn't created or published date work for that. The teaser is used as the summary on the list of issues page, but also as the first paragraph of the table of contents/body. I wonder if just using the existing summary field and changing it to display as part of the regular display, not just list view, is possible and would work better.

**Comment by e-marie-w** [comment]
@quamsta Is the proposed genre field on the article content type something that will be used on Exchanges?

**Comment by e-marie-w** [comment]
One disadvantage of extending the article content type instead of making a new one: figuring out a path alias pattern to cover the news article use case and the publication article use case.

**Comment by joewhitsitt** [comment]
Hmm. Additionally, the article path alias is currently ignored config because of the SiteNow Articles configuration page

Advanced conditions within pathauto: https://www.drupal.org/project/pathauto/issues/3224916

**Comment by joewhitsitt** [comment]
Looks like there are some workarounds mentioned here: https://www.drupal.org/project/pathauto/issues/3126837

We would still have to update docroot/modules/custom/sitenow_articles/src/Form/SettingsForm.php to make sure it doesn't try to re-write the publication article.

Is there a "article type" or something that designates what kind of article it is? That could be tapped into for the alter and used to filter out the batch update.

**Comment by pyrello** [comment]
> Is there a "article type" or something that designates what kind of article it is? That could be tapped into for the alter and used to filter out the batch update.

Would having an issue referenced be enough for this?

**Comment by e-marie-w** [comment]
I can try it!

**Comment by pyrello** [comment]
I'm wondering if the patch at this issue might help? https://www.drupal.org/project/pathauto/issues/3258256. I think it would allow having multiple path alias patterns apply to a single content type and includes a UI to support selecting the path alias pattern for auto-generating the alias. We could potentially use a form alter and form `#states` to check if a value has been set for the issue field and select a value for the path alias pattern field and hide it.

**Comment by e-marie-w** [comment]
So, would my first step be enabling/installing the patch? And then altering code if needed?

**Comment by e-marie-w** [comment]
Update on the patch so far: it lets me add another path alias for the article content type, and I can bulk regenerate the articles to use that pattern type. On the edit form of an article there isn't an option to switch between the two and it defaults to the news/date pattern. Which, maybe is what it is supposed to do? Regardless, we planned for the article pattern to be picked based on if the issue field had a value or not. So that's what I plan to tackle next.

**Comment by joewhitsitt** [comment]
I guess I don't understand the value of the patch. Non-admins don't have access to that bulk generate interface

**Comment by e-marie-w** [comment]
> I guess I don't understand the value of the patch. Non-admins don't have access to that bulk generate interface

I thought it was going to let me switch between pattern options in the edit form. Also, I thought I couldn't add another path alias for a content type that already had one but I'm realizing I never tried and probably should check now.

**Comment by joewhitsitt** [comment]
Maybe the patch is incomplete but yeah, when I first saw this, I was thinking it would be available when editing a node. You can define multiple patterns for the same content type already. It is the order (drag and drop ordering) in which pathauto reads them that determines which pattern to use.

**Comment by pyrello** [comment]
I didn't evaluate the patch for completeness, just saw that it was RTBC. Sorry for the dead end.

**Comment by e-marie-w** [comment]
It was worth checking plus it gave me more experience with using patches.

**Comment by e-marie-w** [comment]
Path alias changing is working (well when hardcoded right now). But I put the function in sitenow_articles.module, and now I'm wondering if that's correct. Because it is changing the articles, but at the same time it's feature-specific...so I'm not sure if it's supposed to go into the split or its current location (with a check that the split is turned on).

I'd try it both ways and see, but the periodical doesn't have a module folder or a .module file right now. And I forgot how that's created -  is there a command to generate it, or do I just create the directory and file myself?

**Comment by joewhitsitt** [comment]
I think this is feature specific and so a feature module should be added to the split so that it only fires if the split is enabled.

there is a `drush generate` command, but can also create the folder/info file/module file yourself.

**Comment by e-marie-w** [comment]
I mentioned in planning that one of the finishing details is finalizing the pathauto alias patterns. Right now I have the publication issue path set to the issue title and for publication articles it's issue_title/article_title. Like /volume-45-issue-2-fall-2022/kitty-cat

The issue titles, at least for Iowa Review and 91st Meridian, end up being some variation of 'Volume # Issue # Season-Year'. I wonder if it makes more sense to programmatically create the title based on the volume, issue, and season input, instead of having a title field they need to fill in. The path could then work like vol/45/issue/2/fall-2022. If down the line we need to have a display of all volumes or all issues part of a specific volume the paths are already set up to work with it. 

The only thing is ideally the article would have this same path, with the addition of article title. Like, vol/45/issue/2/fall-2022/kitty-cat. And I'm not sure that's possible. I can get the issue title because there's a field for an entity reference to it. But I'm not sure I can access the volume and issue numbers without needing to add fields for them. 

I'd love to hear others' thoughts and ideas.

**Comment by pyrello** [comment]
@e-marie-w Thanks for putting some thought into this. 

I'm trying to think through the way that we would break the paths down. I think the main reason for slashes in the URL is to have a higher level collection that the user can find their way to. In the suggested path above, this would lead to:
- `vol` - Would this be a listing of volumes?
- `vol/45` - This seems like the home of volume 45. Presumably this page would also be a listing of issues that belong to volume 45.
- `vol/45/issue` - Not sure what would go here since we already have our list of issues at the previous level up.
- `vol/45/issue/2` - This would be the home of issue 2. This would show our custom table of contents listing articles linked to this issue.
- `vol/45/issue/2/fall-2022` - I'm not sure this needs its own space and is more just an indicator of when the issue was published. Probably this would be combined with the issue title.
- `vol/45/issue/2/fall-2022/kitty-cat` - Article page.

Based on the analysis above, I would likely conclude that we could get away with something like: 
- `volume-45/` - Home of volume 45 including listing of issues that link to this volume.
- `volume-45/fall-2022-issue-2-we-love-cats` - Home of a particular issue in volume 45 including a listing of articles that link to this issue via the table of contents field. Either we would just use the issue title or combine the issue number and title.
- `volume-45/fall-2022-issue-2-we-love-cats/kitty-cat` - Article page.

I forget where we landed on using taxonomy for the volume number. If we did that, then we already have a page for volumes (which will need to have a path alias set if it hasn't already been) and we can go ahead and use that. If we did not create it as a taxonomy reference, then I think we should just include the volume in the title of the issue.

I like the idea of constructing the title from some of the fields. We do this with people already. We would need to add an extra field to hold the issue title, then implement code similar to what we do with people to concatenate the various pieces of the title together. There is also some code to hide the title field on the person node form.

I'd be happy to try to work through how to get the upstream pieces of information in the path.

**Comment by e-marie-w** [comment]
Thanks @pyrello! We do not have a taxonomy for the volume number at this time as neither site uses it beyond having it as part of the issue title. Continuing to just use it as part of the title sounds good.

I forgot we already constructed the title for people, I'll take a look at that and see what I can do.

**Comment by pyrello** [comment]
After looking at #5596 a bit and comparing it with existing solutions, I'm not totally sure about the approach of automatically setting the title. One change I might like to see is renaming the session/year field on the Issue content type to just be called name or title. We could add help text providing an example of using the season and/or year for the name.

**Comment by briand44** [comment]
@pyrello Here are the links to the Law Journal sites that are in progress. Currently they are using the page content type for their journal articles. We are having conversations with them to see if that is the best approach considering their timelines (hoping to launch mid to end of January) or if this work or just using articles and prioritizing the narrow article option makes sense for them. If they stay with pages, I think their plan is to hide the node title (JGRJ is outputting two titles currently because they like the title in the smaller width).

https://lawjgrj.prod.drupal.uiowa.edu/
https://lawilr.prod.drupal.uiowa.edu/

**Comment by e-marie-w** [comment]
> After looking at #5596 a bit and comparing it with existing solutions, I'm not totally sure about the approach of automatically setting the title. 

I can be persuaded. There were a few reasons I did it like this. The first reason was for consistency. I could picture different editors on the same site using different styles - abbreviating with or without punctuation, writing out the entire words, etc. It seems simpler to control that from the start. Another reason was because it felt redundant to capture and display the volume, issue, and season/year fields after they were already listed in the title. But we wanted to capture those fields for schema.org reasons.

If we want to trust our users and let them handle any title differences I can change it. And I can see a possible use case where a publication gives their issues titles instead of identifying by numbers/dates. I'd just have questions about displaying the fields.

> One change I might like to see is renaming the session/year field on the Issue content type to just be called name or title. We could add help text providing an example of using the season and/or year for the name.

That is a great idea! I'll get started on that.

**Comment by pyrello** [comment]
Once we make the change to the session/year field to make it a more generic title field, I think I'm good with rolling with this for now. We may need to make changes down the road, but I think we will be positioned to do so without concern.
